mkosi/kmod.py

```diff
@@ -55,6 +55,39 @@ def filter_kernel_modules(
     return sorted(modules)
 
 
+def filter_firmware(
+    root: Path,
+    firmware: set[Path],
+    *,
+    include: Iterable[str],
+    exclude: Iterable[str],
+) -> list[Path]:
+    firmwared = Path("usr/lib/firmware")
+    with chdir(root):
+        all_firmware = set(firmwared.rglob("*"))
+
+    keep = set()
+    if include:
+        regex = re.compile("|".join(include))
+        for f in all_firmware:
+            rel = os.fspath(Path(*f.parts[3:]))
+            if regex.search(rel):
+                keep.add(rel)
+                firmware.add(f)
+
+    if exclude:
+        remove = set()
+        regex = re.compile("|".join(exclude))
+        for f in firmware:
+            rel = os.fspath(Path(*f.parts[3:]))
+            if rel not in keep and regex.search(rel):
+                remove.add(f)
+
+        firmware -= remove
+
+    return sorted(firmware)
```

Comment:
While this is symmetric with `filter_kernel_modules`, this goes through `resolve_module_dependencies`, which makes `firmware` a set and with this here, we sort it, to directly throw this away again by making it a set at the call site and sorting it again a bit later. You can return `firmware` here and make the function return `set[Path]` instead.